### PR TITLE
transformations: Add convert-arith-to-varith pass

### PR DESCRIPTION
**Pull request by AntonLydike**
Adds a pass `convert-arith-to-varith` that converts arith adds and muls to varith adds and muls. More tests coming.

Stacked on top of #3241

### CODE REVIEW

**Review by n-io — APPROVED** [review]
Thanks, looks good!

File: xdsl/transforms/varith_transformations.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    ("int", "add"): arith.Addi,
+    ("int", "mul"): arith.Muli,
+    ("float", "add"): arith.Addf,
+    ("float", "mul"): arith.Mulf,
+}
+
+
+class MergeVarithOpsPattern(RewritePattern):
+    """
+    Merges operands that are varith or arith ops into
```
Comment (dk949) [inline xdsl/transforms/varith_transformations.py:65]: I'm not quite sure what you mean here

File: xdsl/transforms/varith_transformations.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                new_operands.extend(inp.owner.operands)
+                possibly_erased_ops.append(inp.owner)
+            else:
+                new_operands.append(inp)
+
+        # nothing to do if no new operands are added
+        if len(possibly_erased_ops) == 0:
+            return
+
+        rewriter.replace_matched_op(type(op)(*new_operands))  # pyright: ignore[reportUnknownArgumentType]
```
Comment (dk949) [inline xdsl/transforms/varith_transformations.py:96]: nit: Could you include a comment for why you need to ignore the warning?

**Review by dk949 — APPROVED** [review]

File: xdsl/transforms/varith_transformations.py (hunk 1 above)
Comment (AntonLydike) [reply xdsl/transforms/varith_transformations.py]: Expanded on it!

File: xdsl/transforms/varith_transformations.py (hunk 2 above)
Comment (AntonLydike) [reply xdsl/transforms/varith_transformations.py:127]: Done!

**Pull request merged**